api/v1/api_test.go

```diff
@@ -0,0 +1,273 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package v1
+
+import (
+	"bytes"
+	"encoding/json"
+	"math"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	dto "github.com/prometheus/client_model/go"
+
+	"github.com/go-kit/kit/log"
+	"github.com/golang/protobuf/proto"
+	"github.com/prometheus/pushgateway/storage"
+)
+
+var (
+	logger    = log.NewNopLogger()
+	testFlags = map[string]string{
+		"flag1": "value1",
+		"flag2": "value2",
+		"flag3": "value3",
+	}
+	testBuildInfo = map[string]string{
+		"build1": "value1",
+		"build2": "value2",
+		"build3": "value3",
+	}
+
+	mf1 = &dto.MetricFamily{
+		Name: proto.String("mf1"),
+		Type: dto.MetricType_UNTYPED.Enum(),
+		Metric: []*dto.Metric{
+			{
+				Label: []*dto.LabelPair{
+					{
+						Name:  proto.String("instance"),
+						Value: proto.String("instance2"),
+					},
+					{
+						Name:  proto.String("job"),
+						Value: proto.String("job1"),
+					},
+				},
+				Untyped: &dto.Untyped{
+					Value: proto.Float64(-3e3),
+				},
+			},
+		},
+	}
+
+	mf2 = &dto.MetricFamily{
+		Name: proto.String("mf2"),
+		Help: proto.String("doc string 2"),
+		Type: dto.MetricType_GAUGE.Enum(),
+		Metric: []*dto.Metric{
+			{
+				Label: []*dto.LabelPair{
+					{
+						Name:  proto.String("basename"),
+						Value: proto.String("basevalue2"),
+					},
+					{
+						Name:  proto.String("instance"),
+						Value: proto.String("instance2"),
+					},
+					{
+						Name:  proto.String("job"),
+						Value: proto.String("job1"),
+					},
+					{
+						Name:  proto.String("labelname"),
+						Value: proto.String("val2"),
+					},
+				},
+				Gauge: &dto.Gauge{
+					Value: proto.Float64(math.Inf(+1)),
+				},
+			},
+			{
+				Label: []*dto.LabelPair{
+					{
+						Name:  proto.String("instance"),
+						Value: proto.String("instance2"),
+					},
+					{
+						Name:  proto.String("job"),
+						Value: proto.String("job1"),
+					},
+					{
+						Name:  proto.String("labelname"),
+						Value: proto.String("val1"),
+					},
+				},
+				Gauge: &dto.Gauge{
+					Value: proto.Float64(math.Inf(-1)),
+				},
+			},
+		},
+	}
+
+	grouping1 = map[string]string{
+		"job":      "job1",
+		"instance": "instance1",
+	}
+)
+
+func templateMetrics(metricStore storage.MetricStore) []interface{} {
+	familyMaps := metricStore.GetMetricFamiliesMap()
+	res := make([]interface{}, 0)
+	for _, v := range familyMaps {
+		metricResponse := make(map[string]interface{})
+		metricResponse["labels"] = v.Labels
+		metricResponse["last_push_successful"] = v.LastPushSuccess()
+		for name, metricValues := range v.Metrics {
+			metricFamily := metricValues.GetMetricFamily()
+			uniqueMetrics := metrics{
+				Type:      metricFamily.GetType().String(),
+				Help:      metricFamily.GetHelp(),
+				Timestamp: metricValues.Timestamp,
+				Metrics:   makeEncodableMetrics(metricFamily.GetMetric(), metricFamily.GetType()),
```

Comment:
The problem with this approach is that you are calling `makeEncodableMetrics` to create the expected test result. You are not really testing that part, you are essentially creating a self-fulfilling prophecy: Calling the same function in the test creates the same results.

That's why I would go for an approach where you just hardcode a "golden" JSON string and compare it directly.